Cython/Compiler/Optimize.py

```diff
@@ -1417,18 +1417,33 @@ def visit_PrimaryCmpNode(self, node):
         else:
             return node
 
-        if not isinstance(node.operand2, (ExprNodes.TupleNode,
+        if not (isinstance(node.operand2, (ExprNodes.TupleNode,
                                           ExprNodes.ListNode,
-                                          ExprNodes.SetNode)):
+                                          ExprNodes.SetNode)) or node.operand2.type.is_ctuple):
             return node
 
         lhs = node.operand1
-        args = node.operand2.args
+
+        if node.operand2.type.is_ctuple and not isinstance(node.operand2, ExprNodes.TupleNode):
+            rhs = UtilNodes.ResultRefNode(node.operand2)
+            args = [ExprNodes.IndexNode(node.pos,
+                        base=rhs,
+                        index=ExprNodes.IntNode(node.pos,
+                            value=str(index),
+                            constant_result=index,
+                            type=PyrexTypes.c_py_ssize_t_type)
+                        ).analyse_types(self.current_env()).coerce_to(node.operand1.type, self.current_env())
+                    for index, type_ in enumerate(rhs.type.components)
+                        if node.operand1.type.assignable_from(type_) or \
+                            ((node.operand1.type.is_pyobject and PyrexTypes.py_object_type.assignable_from(type_)) or \
+                             (type_.is_pyobject and PyrexTypes.py_object_type.assignable_from(node.operand1.type)))]
```

Comment:
If you need the type information here, why not do this in `PrimaryCmpNode.analyse_types()`? This code is just spelling out how we need to implement `in` tests on ctuples correctly. There are cases where they don't currently work at all, e.g. for ctuples of pointers which cannot be converted to Python tuples. Thus, it's not actually an optimisation but rather a language feature implementation. It does not need to happen in the (optimising) `FlattenInListTransform` and `analyse_types()` is a good place for this kind of syntax tree replacement.